src/pudl/transform/eia861.py

```diff
@@ -731,6 +731,72 @@ def advanced_metering_infrastructure(tfr_dfs):
         dict: A dictionary of transformed EIA 861 dataframes, keyed by table name.
 
     """
+    idx_cols = [
+        "utility_id_eia",
+        "state",
+        "balancing_authority_code_eia",
+        "report_date",
+    ]
+
+    ###########################################################################
+    # Tidy Data:
+    ###########################################################################
+    logger.info("Tidying the EIA 861 Advanced Metering Infrastructure table.")
+    # Clean up values just enough to use primary key columns as a multi-index:
+    # Transform values with fix_eia_na()
+    logger.debug(
+        "Cleaning up EIA861 Advanced Metering Infrastructure index columns so we can tidy data.")
+    clean_ami = (
+        tfr_dfs["advanced_metering_infrastructure_eia861"].copy()
+        .assign(balancing_authority_code_eia=lambda x: x.balancing_authority_code_eia.fillna("UNK"))
+        .dropna(subset=["utility_id_eia"])
+        .query("utility_id_eia not in (88888, 99999)")
+        .astype({"utility_id_eia": pd.Int64Dtype()})
+        .set_index(idx_cols)
+    )
+
+    # Split the table into index, data, and "denormalized" columns for processing:
+    # Separate customer classes and reported data into a hierarchical index
+    logger.debug(
+        "Stacking EIA861 Advanced Metering Infrastructure data columns by customer class.")
+    data_cols = _filter_customer_cols(clean_ami, CUSTOMER_CLASSES)
+    data_cols.columns = (
+        data_cols.columns.str.split("_", n=1, expand=True)
+        .set_names(["customer_class", None])
+    )
+
+    # Now stack the customer classes into their own categorical column,
+    data_cols = (
+        data_cols.stack(level=0, dropna=False)
+        .reset_index()
+    )
+    denorm_cols = (
+        _filter_non_customer_cols(clean_ami, CUSTOMER_CLASSES)
+        .reset_index()
+    )
+
+    # Merge the index, data, and denormalized columns back together
+    tidy_ami = pd.merge(denorm_cols, data_cols, on=idx_cols)
+
+    # Remove the now redundant "Total" records -- they can be reconstructed
+    # from the other customer classes.
+    tidy_ami = tidy_ami.query("customer_class!='total'")
```

Comment:
If you notice that we're ending up doing exactly this set of operations for multiple tables, we should probably turn it into a local `_helper_function()` of some kind.

Reply:
Will do